contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -86,23 +27,61 @@ pub fn route_to_router(
     Ok(Response::new().add_messages(router.route(msgs)))
 }
 
-pub fn route_to_nexus(
+pub fn route_message_with_token_to_nexus(
     storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    info: MessageInfo,
+    msg: router_api::Message,
+) -> Result<Response<nexus::execute::Message>> {
+    let token = info
+        .token()
+        .change_context(Error::InvalidToken)?
+        .ok_or(report!(Error::InvalidToken))?;
+
+    route_message_to_nexus(storage, querier, msg, Some(token))?
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Result::Ok)
+}
+
+pub fn route_messages_to_nexus(
+    storage: &mut dyn Storage,
+    querier: QuerierWrapper,
     msgs: Vec<router_api::Message>,
 ) -> Result<Response<nexus::execute::Message>> {
-    let msgs = msgs
-        .into_iter()
-        .filter_map(|msg| match state::is_message_routed(storage, &msg.cc_id) {
-            Ok(true) => None,
-            Ok(false) => Some(Ok(msg)),
-            Err(err) => Some(Err(err)),
-        })
-        .collect::<Result<Vec<_>>>()?;
+    msgs.into_iter()
+        .map(|msg| route_message_to_nexus(storage, querier, msg, None))
+        .collect::<Result<Vec<_>>>()?
+        .then(|msgs| msgs.concat())
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Result::Ok)
+}
 
-    msgs.iter()
-        .try_for_each(|msg| state::set_message_routed(storage, &msg.cc_id))?;
+fn route_message_to_nexus(
+    storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    msg: router_api::Message,
+    token: Option<Coin>,
+) -> Result<Vec<CosmosMsg<nexus::execute::Message>>> {
+    if state::is_message_routed(storage, &msg.cc_id)? {
+        return Ok(vec![]);
+    }
+
+    state::set_message_routed(storage, &msg.cc_id)?;
 
-    let msgs: Vec<nexus::execute::Message> = msgs.into_iter().map(Into::into).collect();
+    let client: nexus::Client = client::CosmosClient::new(querier).into();
+    let config = load_config(storage)?;
 
-    Ok(Response::new().add_messages(msgs))
+    let mut msg: nexus::execute::Message = msg.into();
+    msg.token.clone_from(&token);
```

Comment:
Ah, yes, forgot about this. I'd suggest implementing a `impl From<(router_api::Message, Option<Coin>) for nexus::execute::Message` instead to make this a bit safer. `route_message_to_nexus` could then take in just `nexus::execute::Message` (and caller handles the BankMsg) to make it cleaner perhaps?

Reply:
I don't like the idea of "caller handles the BankMsg". It can easily become a future bug to me cuz one may not realize they are a package of one thing, especially now even the order matters. I don't hate the From trait though.

Reply:
Yeah, I agree with the trade off of the caller handling it. Since the caller is in this module itself, I was ok with it to reduce the complexity. And yeah, From trait would be good to have still